File: 1482.py
class Solution:
    def minDays(self, bloomDay: List[int], m: int, k: int) -> int:
        n = len(bloomDay)
        if n < m * k:
            return -1
        l = min(bloomDay)
        r = max(bloomDay) + 1
        while l < r:
            p = (l + r) // 2
            bonquet = 0
            segCount = 0
            for x in bloomDay:
                if x <= p:
                    segCount += 1
                else:
                    bonquet += segCount // k
                    segCount = 0
                if bonquet >= m:
                    break
            bonquet += segCount // k
            if bonquet < m:
                l = p + 1
            else:
                r = p
        return l

File: sol/1482/1482.cpp
struct Solution {
    inline int minDays(vector<int>& bloomDay, int m, int k) {
        int n = bloomDay.size();
        if (n / k < m){
            return -1;
        }
        auto [lp, rp] = minmax_element(bloomDay.begin(), bloomDay.end());
        int l = *lp, r = *rp + 1, p, bonq, seg;
        while (l < r){
            p = (l + r) / 2;
            bonq = 0;
            seg = 0;
            for (int &x : bloomDay){
                if (x <= p){
                    ++seg;
                } else {
                    bonq += seg / k;
                    seg = 0;
                }
                if (bonq >= m){
                    break;
                }
            }
            bonq += seg / k;
            if (bonq < m){
                l = p + 1;
            } else {
                r = p;
            }
        }
        return l;
    }
};
